lib/cea/mp4_cea_parser.js

```
@@ -0,0 +1,229 @@
+
+    new Mp4Parser()
+        .box('moof', Mp4Parser.children)
+        .box('traf', Mp4Parser.children)
+        .fullBox('trun', (box) => {
+          goog.asserts.assert(
+              box.version != null && box.flags!=null,
+              'TRUN is a full box and should have a valid version & flags.');
+
+          const parsedTRUN = shaka.util.Mp4BoxParsers.parseTRUN(
+              box.reader, box.version, box.flags);
+
+          sampleData = parsedTRUN.sampleData;
+        })
+
+        .fullBox('tfhd', (box) => {
+          goog.asserts.assert(
+              box.flags != null,
+              'TFHD is a full box and should have valid flags.');
+
+          const parsedTFHD = shaka.util.Mp4BoxParsers.parseTFHD(
+              box.reader, box.flags);
+
+          // If specified, defaultSampleDuration and defaultSampleSize
+          // override the ones specified in the TREX box
+          defaultSampleDuration = parsedTFHD.defaultSampleDuration
+            || this.defaultSampleDuration_;
+
+          defaultSampleSize = parsedTFHD.defaultSampleSize
+            || this.defaultSampleSize_;
+
+          const trackId = parsedTFHD.trackId;
+
+          // Get the timescale from the track Id
+          if (this.trackIdToTimescale_.has(trackId)) {
+            timescale = this.trackIdToTimescale_.get(trackId);
+          }
+        })
+
+        .fullBox('tfdt', (box) => {
+          goog.asserts.assert(
+              box.version != null,
+              'TFDT is a full box and should have a valid version.');
+
+          const parsedTFDT = shaka.util.Mp4BoxParsers.parseTFDT(
+              box.reader, box.version);
+
+          baseMediaDecodeTime = parsedTFDT.baseMediaDecodeTime;
+        })
+        .box('mdat', (box) => {
+          if (baseMediaDecodeTime === null) {
+            // This field should have been populated by
+            // the Base Media Decode time in the TFDT box
+            return;
+          }
```

Perhaps this would be another good use of the INVALID_MP4_CEA error you've introduced.

---

If I throw an error here, would it interrupt video playback? I don't want to potentially stop video playback just because we received one bad box in a media fragment.

---

Yes, I expect it would stop playback.  If this error caused `appendBuffer` to fail in `MediaSourceEngine`, then that would propagate up to `StreamingEngine` and it would stop streaming.  But the same is true of the other place you throw INVALID_MP4_CEA when parsing the init segment.

I would suggest that we throw on invalid input and let the caller (`MediaSourceEngine`, I think, in this case) decide what errors to catch and ignore.  It might make sense to log this error there and potentially skip some captions without letting it propagate up to `StreamingEngine` and cause a playback failure.

But by throwing on invalid input, we leave that decision to the caller, which I would argue makes this a more useful tool if it's ever extracted to another context.